### PR TITLE
Celo/cdf 21249/attempt number metrics

### PR DESCRIPTION
**Pull request by mr-celo**
### Add attempt number to metric names

By adding the attempt number to the metrics it will be possible to get
the highest attempt for each stage and task and allowing ignoring
metrics coming from failed/retried tasks. A good example would be the
effective number of rows added.

Though, the metrics don't allow labels, and instead uses dot-delimited
fields such as job id.
By keeping always a consistent number of fields, these can be extracted
while aggregating the metrics. Therefore, it is the responsibility of
the MetricSource to enforce this consistency.

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 1]
```diff
@@ -43,9 +62,53 @@ class MetricsSource {
...
+    metricsMap.remove(key)
+    SparkEnv.get.metricsSystem.removeSource(source)
   }
 }
 
 // Singleton to make sure each metric is only registered once.
 object MetricsSource extends MetricsSource
+
+class ExpiringCounter(key: String, source: Source) extends Counter {
+  private var (deathTimer, deathTimerTask): (Timer, TimerTask) = createTimer
```
Comment (mr-celo) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:81]: Unsure if the timer will survive long enough to deregister the metric upon task completion.
This will also not work on a crash of the executor, and the metrics will linger.

Another solution would be to move this to the driver, but the instance of the MetricSource (so that it can be removed from the MetricsSystem) gets a little more complicated. Looking into this approach now, though

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 2]
```diff
@@ -2,22 +2,39 @@ package org.apache.spark.datasource
...
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
+    val key = s"$metricNamespace.$metricName"
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
     val wrapped = Eval.later {
       val counter = new Counter
-      registerMetricSource(metricNamespace, metricName, counter)
-      counter
+      val source = registerMetricSource(metricNamespace, metricName, counter)
+      new ExpiringCounter(key, source)
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:35]: feels like it may start timer even if entry is already present, then unadded value will unregister prematurely. Could be that Eval.later introduces some laziness that prevents this, but sounds fragile even if it does

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 3]
```diff
@@ -2,22 +2,39 @@ package org.apache.spark.datasource
...
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:29]: it'd change metric name? Then we'll lose metrics under old names?

Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:29]: It will. This will be coupled with driver changes

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 2 above)
Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:35]: That is fair, I'd say. The registering of a new metric with the same name can be considered as the metric being "active". Therefore, it will reset the inactivity timer.
Probably fairer than receiving a metric that will die moments later

Or did I misunderstand your concern?

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 4]
```diff
@@ -43,9 +62,53 @@ class MetricsSource {
...
+    deathTimerTask.cancel()
+    deathTimer.cancel()
+
+    val (timer, task) = createTimer
+    deathTimer = timer
+    deathTimerTask = task
+  }
+
+  override def inc(n: Long): Unit = {
+    resetTimer()
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:105]: some potential for perf degradation here, before it was more or less simple increment, now it is noticeable work per increment (a little less scary since it'd probably be batch-increment but still). Especially the recreation of Timer and its thread

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 3 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:29]: Splitting the PR in 2 (expiration and per-stage breakdown) + adding feature flags for those would make sense as both changes have their own risks (perf or breaking metrics somehow and doing something to TaskContext that isn't allowed in executors + if anything depended on old metric names)

Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:29]: I broke it down like that in terms of commits, for the same reason.
But I coupled them in a single PR because the number of metrics will increase significantly. So, by not having them together, a new bigger perf risk arises with the great increase of metrics.

But your point is valid. If something on the driver can address the stale metrics (what I'm looking into now), then both should be paired together within a deployment.

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 2 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:35]: If 2 getOrCreateCounter calls create and start 2 counters, only one ends up in the map and receive events, the other will decide nothing happens and will deregister both by metric name

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 5]
```diff
@@ -43,9 +62,53 @@ class MetricsSource {
...
+class ExpiringCounter(key: String, source: Source) extends Counter {
+  private var (deathTimer, deathTimerTask): (Timer, TimerTask) = createTimer
+
+  private def createTimer = {
+    val task = new TimerTask {
+      override def run =
+        MetricsSource.deregisterMetricSource(key, source)
+    }
+
+    val timer = new Timer(key)
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:89]: here's a thread leak here. Even when we don't have references to TimerThread it will still be considered a garbage collection root by being an alive/running thread

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:89]: Or at least it seems so. There's some code in Timer class that tries to stop the thread when Timer becomes garbage-collectible which might be enough to let the thread stop and be collected, but a bit hard to be really sure this chain would work and no other references path keeps it all alive

File: build.sbt [hunk 6]
```diff
@@ -46,7 +46,13 @@ lazy val commonSettings = Seq(
...
   licenses := List("Apache 2" -> new URL("http://www.apache.org/licenses/LICENSE-2.0.txt")),
   homepage := Some(url("https://github.com/cognitedata/cdp-spark-datasource")),
-  scalacOptions ++= Seq("-Xlint:unused", "-language:higherKinds", "-deprecation", "-feature"),
+  scalacOptions ++= Seq("-Xlint:unused", "-language:higherKinds", "-deprecation", "-feature") ++ (CrossVersion.partialVersion(scalaVersion.value) match {
+    // We use JavaConverters to remain backwards compatible with Scala 2.12,
+    // and to avoid a dependency on scala-collection-compat
+    case Some((2, 13)) => Seq("-Wconf:src=src/main/scala/cognite/spark/v1/MetricsSource.scala&cat=deprecation:i")
+    case _ => Seq.empty
+  }),
+  // scalacOptions ++= Seq("-Xlint:unused", "-language:higherKinds", "-deprecation", "-feature", "-Wconf:src=src/main/scala/cognite/spark/v1/MetricsSource.scala&cat=deprecation:i"),
```
Comment (dmivankov) [inline build.sbt:55]:
```suggestion
```

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 3 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:28]: What was the conclusion of doing some of the logic in the driver?

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 7]
```diff
@@ -2,24 +2,41 @@ package org.apache.spark.datasource
...
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:28]: On the cardinality front, this can make x200 (raw partitions), x3-5 (task attempts), x3-5 (stage attempts) -> x?? (num stages). And when it gets propagated to prometheus (in driver it does) it still impacts prometheus even after it stops being exported (prometheus still remembers names+labels for the retention period).

Would be good to
- keep old metric for transition period
- for new metrics add a simple prefix so that driver could easily ignore them when exporting to prometheus if it decides to
- add feature flag that allows reducing cardinality back (disable extended metrics completely), increase could be causing too big change in load towards metrics store, potentially, would be good to have a quick switch off
- maybe add feature flags per dimension to disable the breakdown: partition (very deep debugging thing and big multiplier at the same time, first candidate for being disabled), maybe others

Also maybe it is still to difficult to correctly discard metrics for failed/lost and retried data reads/writes, maybe there's some better way to get number of items read/write. Otoh metrics cleanup for completed jobs is a pretty clear improvement which would be nice to get done sooner than retry-aware counting (and so independently of it as it is proving to be not that trivial to get those numbers..)

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:28]: Following some discussion I'll need to read into what exactly happens with those metrics :eyes: plus when the driver side is ready would be easier to review them together :+1:

Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala]: * The new driver is backwards compatible, so we can bring this change in an out
* The driver will only send an extra label to prometheus: `completed`

The driver PR is ready: https://github.com/cognitedata/jetfire-backend/pull/2285

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 8]
```diff
@@ -32,20 +49,64 @@ class MetricsSource {
...
+    source
+  }
+
+  /**
+    * Remove the metrics from a job.
+    *
+    * This method will deregister the metric from Spark's org.apache.spark.metrics.MetricsSystem
+    * and stops tracking that it was published
+    */
+  def removeJobMetrics(metricNamespace: String, jobId: String): Unit = {
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:72]:
```suggestion
  def unregisterJobMetrics(metricNamespace: String, jobId: String): Unit = {
```
nit: maybe something like that to reflect that it's mostly about in-mem metrics

File: src/main/scala/cognite/spark/v1/CdfRelation.scala [hunk 9]
```diff
@@ -12,21 +12,19 @@ abstract class CdfRelation(config: RelationConfig, shortNameStr: String)
...
-  @transient lazy protected val itemsCreated: Counter =
-    MetricsSource.getOrCreateCounter(config.metricsPrefix, s"$shortName.created")
-  @transient lazy protected val itemsUpdated: Counter =
-    MetricsSource.getOrCreateCounter(config.metricsPrefix, s"$shortName.updated")
-  @transient lazy protected val itemsDeleted: Counter =
-    MetricsSource.getOrCreateCounter(config.metricsPrefix, s"$shortName.deleted")
+  protected def itemsRead: Counter = getOrCreateCounter("read")
+  protected def itemsCreated: Counter = getOrCreateCounter("created")
+  protected def itemsUpdated: Counter = getOrCreateCounter("updated")
+  protected def itemsDeleted: Counter = getOrCreateCounter("deleted")
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/CdfRelation.scala:18]:
```suggestion
  @transient lazy protected val itemsRead: Counter = getOrCreateCounter("read")
  @transient lazy protected val itemsCreated: Counter = getOrCreateCounter("created")
  @transient lazy protected val itemsUpdated: Counter = getOrCreateCounter("updated")
  @transient lazy protected val itemsDeleted: Counter = getOrCreateCounter("deleted")
```
we can still keep it being a field?

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 10]
```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
...
+import org.apache.spark._
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
   val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:14]: is that TaskContext coming from a thread-local horror land? :thinking:

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 11]
```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
...
 import java.util.concurrent.ConcurrentHashMap
+import org.apache.spark._
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
   val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:13]:
```suggestion
  def getOrCreateDetailedCounter(metricNamespace: String, name: String): Counter = {
```
or something like that to indicate that meaning changes a lot while type signature remains the same `(String, String)->Counter`, this would enforce fixing references to it here or for external consumers

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 12]
```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
...
+    val ctx = Option(TaskContext.get())
+
+    def getNumberOrEmpty(getter: TaskContext => AnyVal): String =
+      ctx.map(getter).map(_.toString()).getOrElse("")
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:17]:
```suggestion
    def getNumberOrEmpty(getter: TaskContext => String): String =
      ctx.map(getter).getOrElse("")
```
to be more type-safe, at call-sites just add a quick .toString()

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 10 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:14]: if it is then val->def change makes sense but would be better than to move that stuff around, let MetricSource.getCounter take TaskContext as parameter and pass it in CdfRelation and even there maybe also take it as a parameter and require filling it at metric update sites like
```scala
        if (config.collectMetrics) {
          itemsRead.inc()
        }
```
so we definitely have more control on which thread it gets executed on (datasource repo uses cats-effect in various places, to switches between threads are quite likely)

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 12 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:17]: Not sure if there's a common numeric type that we could use to make sure getter gets a (non-floating point) number (Long?)

File: src/main/scala/cognite/spark/v1/MetricsSource.scala [hunk 13]
```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
...
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"<$stageId>[$stageAttempt]#<$partitionId>[$taskAttempt]#$name"
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/MetricsSource.scala:26]: for a big format like that it might be nice to put the reading part also in this lib + a test or two to illustrate the format (normal case, case with empty strings). Also maybe since fields are so many a small data type to represent taskContext+name

Only a little surprised that current tests didn't notice much of a change. But to hypothesize, if a consumer of old metric format sees it it will interpret the whole 5-part thing as "metricName"? Might be especially interesting how they interact in case metrics are persisted and either old or new reader might see both kinds in transition periods.

Also had a look at https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md - exposition format probably doesn't matter a lot as it should encode, but https://prometheus.io/docs/concepts/data_model/ metric name restrictions might be important as not sure what will happen when disallowed prometheus-metric-name characters are used (would they be replaced, converted, or will something fail?). Not seeing an explicit length restriction there. Would feel more safe if we stay within allowed prometheus names, so maybe some variation on `::`? Like `":stage_:$stageId:stagetry_:$stageAttempt:part_:$partitionId:tasktry_:$taskAttempt:$name" ?`

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 10 above)
Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:14]: Got it. Yeah, I suppose that can be done. Then the counters can become fields again (referring to the comments above 👆 )

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 13 above)
Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:26]: Good point. Will change to not risk having the name change from underneath our feet.
As for the tests, the brunt of it happened it the repo that uses this as a datasource

Will also bring the parser here.

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 11 above)
Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:13]: Refactoring time

File: src/main/scala/cognite/spark/v1/MetricsSource.scala (hunk 12 above)
Comment (mr-celo) [reply src/main/scala/cognite/spark/v1/MetricsSource.scala:17]: `Long` should cover them all, yeah 👍 Especially bc this is a local function

File: src/test/scala/cognite/spark/v1/MetricsSourceTest.scala [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      }
+    }
+  }
+
+  it should "work with old name format" in {
+    val name = "my.name:goes.here"
+
+    name match {
+      case MetricsSource.metricNameRegex(stageId, stageAttempt, partitionId, taskAttempt, name) => {
+        assert(stageId === null)
```
Comment (Jacob-Eliat-Eliat) [inline src/test/scala/cognite/spark/v1/MetricsSourceTest.scala:48]: Ah I get why it's like this but that's kind of sad :s Have you considered writing a class that would contain the regex and these value and would put them in options with value None? Might not be worth the effort though, but null in scala is meh

**Review by Jacob-Eliat-Eliat — APPROVED** [review]
lgtm obviously the lack of null handling by scala's regex is a bit sad but that's a bit of a nitpick since it's java seeping through

File: src/test/scala/cognite/spark/v1/MetricsSourceTest.scala (hunk 14 above)
Comment (mr-celo) [reply src/test/scala/cognite/spark/v1/MetricsSourceTest.scala]: Wrapped and handled empty strings as well
